### DISCUSSION TITLE
Dialogs

### DISCUSSION THREAD

**Discussion by Aytharn** [fsprojects/Avalonia.FuncUI#372] [Q&A]
How can I show dialogs in Avalonia, and use their result?

**Comment by Numpsy** [comment]
I have't tried FuncUI with distinct dialog windows (file pickers etc aside), but fwiw I have used https://github.com/AvaloniaUtils/DialogHost.Avalonia for a few simple things and that seems to work ok.

There is https://github.com/mysteryx93/HanumanInstitute.MvvmDialogs but I've never tried to use it.

**Reply by Aytharn** [reply]
DialogHost I cannot figure out how to use it:

```
let dlg = DialogHostAvalonia.DialogHost()
dlg.CloseOnClickAway <- true
dlg.IsOpen <- true
dlg.DialogContent <- TextBlock.create [
    TextBlock.text (getLabel "Hello")
]
```

compiles but doesn't do anything, where can I find a sample or a little doc?

**Reply by Numpsy** [reply]
Think the DialogHost needs to be attached to the parent window itself to work, There is a demo app in its own repo under https://github.com/AvaloniaUtils/DialogHost.Avalonia/tree/main/DialogHost.Demo but I can see about making a self contained FuncUI example later

**Reply by Aytharn** [reply]
It would much appreciated, that and perhaps animations as well, I found very hard to understand how to create simple things like Transitions and Dialogs in FuncUI, it's a shame. Other than that is an excellent alternative to SwiftUI

**Reply by JaggerJo** [reply]
I usually write a small modal system per app I create 😅

Dialogs in general should be simple as well.

For animations the only thing missing is documentation and syntactic sugar.

Time is the limiting factor on my end. Feel free to tackle any of the issues above.

**Comment by AngelMunoz** [comment]
I'm not sure what kind of dialogs do you mean, do you mean system dialogs (like file pickers) or your own windows?

For your own windows I have an old sample for one approach that shares a observable to elmish subscriptions so both windows know what each other are talking about:
- https://github.com/AngelMunoz/MultiWindow

In the more recent "Component" version should be simpler to implement

For System Dialogs I have another slightly more up to date version:
- https://github.com/AngelMunoz/WindowSample

**Comment by Aytharn** [comment]
Sorry guys, none of your answers seems to lead anywhere, nor to be related
I will try to rephrase: 
How to pop up a Dialog window made in FuncUI, not in xaml, just FuncUI, plain full, and not a system one (no file picker no nothing a custom one: displaying a text, a stackpanel or whatever else I like) .
I cannot find any documentation or sample about this, I would say it is a quite important feature to be explained, isn't it? ;-)

**Reply by JaggerJo** [reply]
Approach 1: Create a new window and use it as a dialog

At any point you can just create a new window in F# and show it. This will work on all desktop platforms, but not on mobile as there is no such thing as multiple windows when dealing with a `ISingleViewLifetime`. It's possible to make the window a child of the current one and get some modal behaviour IIRC.

Approach 2: In app modal overlay

Implement your own logic to show modals of any kind. You can do this by having a `Panel` view that contains your normal apps content + a modal host that shows a stack of modal views. 

You can then create a nice abstraction for this, like a modal service, ... 

I'd go with the latter.

**Reply by Aytharn** [reply]
The latter is similar to how my app is working indeed, yet I don't want to totally hide or loose the state of the running view, It would be nice to emulate the popup in someway similar to a SwiftUI ZPanel.

Would that be possible? I need the background no more sensible to any interactivity then...

**Reply by Aytharn** [reply]
Tried, Solution 2 won't work or I will loose nice scrolls since I need to put stuff inside grid to make fluid scrolling work.
So neither of the two solutions seems a way to go.
I see now why there are no Samples App :-p

**Reply by Numpsy** [reply]
Hmm, I forgot that I'd said I'd have a go at an example, but DialogHost.Avalonia works along the lines of approach 2 - you create a host control, have your own content as a child of that, and then when you display the 'dialog' it appears as an overlay on top of the other content (which becomes disabled in the background).
I've only tried it with a single dialog in the app and my main reason for that was trying to use it in web assembly where you can't really do multiple windows, but I thought it was supposed to support multiple dialogs.

**Reply by JaggerJo** [reply]
<img width="512" alt="Screenshot 2024-01-06 at 17 32 38" src="https://github.com/fsprojects/Avalonia.FuncUI/assets/13090415/30d11f20-269a-4c44-b808-0b398921a86d">

Added an example in the Todo app. 

Should work with version 1.2.0 of FuncUI. (just published, added some quality of life improvements)

https://github.com/fsprojects/Avalonia.FuncUI/blob/master/src/Examples/Component%20Examples/Examples.TodoApp/ModalHost.fs

**Reply by Aytharn** [reply]
Cool! But Unfortunately I don't see 1.2.0 released...

**Reply by JaggerJo** [reply]
They are available now.

**Comment by Aytharn** [comment]
Ok, it works this way, but there's a caveat: it's better to handle the dialog in a separate component than the background, to avoid unwanted state changes with complex apps
